core/src/main/java/com/linecorp/armeria/client/retry/RetryingClient.java

```diff
@@ -133,43 +136,71 @@ protected final long getNextDelay(ClientRequestContext ctx, Backoff backoff, lon
-            return -1;
+        final long responseTimeoutMillis = state.responseTimeoutMillis();
+        if (responseTimeoutMillis == 0) { // timeout is disabled.
+            return nextDelay;
         }
 
-        final long responseTimeoutMillis = TimeUnit.NANOSECONDS.toMillis(deadlineNanos - System.nanoTime());
-        if (responseTimeoutMillis > 0) {
-            return responseTimeoutMillis;
+        if (responseTimeoutMillis < 0) {
+            // make the client retry as soon as possible, so ResponseTimeoutException occurs when
+            // resetResponseTimeout(ctx) is invoked.
+            return 0;
         }
 
-        // 0 is not from the first, but subtracted to that, so it's timeout.
-        throw ResponseTimeoutException.get();
+        return Math.min(nextDelay, responseTimeoutMillis);
     }
 
     private static class State {
 
         private final int defaultMaxAttempts;
+        private final long responseTimeoutMillis;
+
+        final long responseTimeoutMillisOfEachReq;
 
         private Backoff lastBackoff;
         private int currentAttemptNoWithLastBackoff;
         private int totalAttemptNo;
+
         long deadlineNanos;
+        boolean responseTimeoutOccurred;
 
-        State(int defaultMaxAttempts, long responseTimeoutMillis) {
-            init(responseTimeoutMillis);
+        State(int defaultMaxAttempts, long responseTimeoutMillisOfEachReq, long responseTimeoutMillis) {
             this.defaultMaxAttempts = defaultMaxAttempts;
+            this.responseTimeoutMillisOfEachReq = responseTimeoutMillisOfEachReq;
+            this.responseTimeoutMillis = responseTimeoutMillis;
+            totalAttemptNo = 1;
+            resetDeadline();
         }
 
-        void init(long responseTimeoutMillis) {
-            setDeadlineOfThisRequest(responseTimeoutMillis);
-            lastBackoff = null;
-            totalAttemptNo = 1;
+        void resetDeadline() {
+            deadlineNanos = responseTimeoutMillis <= 0 ?
+                            0 : System.nanoTime() + TimeUnit.MILLISECONDS.toNanos(responseTimeoutMillis);
+            responseTimeoutOccurred = false;
+        }
+
+        /**
+         * Returns the smaller value between {@code responseTimeoutMillisOfEachReq} and
+         * remaining {@code responseTimeoutMillis}.
+         *
+         * @return 0 if the response timeout for both of each request and whole retry is disabled or
+         *         -1 if the elapsed time from the first request has passed {@code responseTimeoutMillis}
+         */
+        long responseTimeoutMillis() {
+            if (deadlineNanos == 0) { // response timeout to the whole is disabled.
+                return responseTimeoutMillisOfEachReq;
+            }
+
+            final long responseTimeoutMillis = TimeUnit.NANOSECONDS.toMillis(deadlineNanos - System.nanoTime());
+
+            if (responseTimeoutMillis <= 0) { // 0 is not from the first, but subtracted to that, it's timeout.
```

Comment:
I'm not sure I understood the comment. Could you rephrase it a little bit?

Reply:
Rephrased!